slippymap/cachedtileprovider.go

```diff
@@ -0,0 +1,111 @@
+package slippymap
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+)
+
+// TileProvider generates URLs (either https:// or file://) to the tile for the coord
+type TileProvider interface {
+	GetTileAddress(x, y, z int) (tilePath string, err error)
+}
+
+func NewCachedTileProvider(tileCachePath string, tileProvider TileProvider) *CachedTileProvider {
```

Comment:
I don't know what's going on here... :-)

Might need you to explain at some point.

Reply:
Using interfaces so we can transparently swap out the caching vs not-caching implementation in wasm as necessary.

`TileProvider` -> something that generates tile addresses/URLs/whatever

`OSMTileProvider` -> generates the real OSM CDN addresses

`CachedTileProvider` -> generates on-disk URLs for cached files. It uses another TileProvider (OSMTileProvider) to generate the URLs to download in order to cache it.

Maybe the code would be clearer if I made the signature:

```
 func NewCachedTileProvider(tileCachePath string, tileProvider *OSMTileProvider) *CachedTileProvider 
```